text/0715-implicit-application-query-params.md

```diff
@@ -0,0 +1,96 @@
+---
+Stage: Accepted
+Start Date: 2020-01-30
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/715
+---
+
+<!---
+Directions for above:
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Implicit Application Query Params
+
+## Summary
+
+This RFC doesn't aim to change existing behavior but does aim to allow opting
+developers to use *any* Query Param passed to `transitionTo`. This will hopefully
+improve the ergonomics of using query params in most default situations.
+
+## Motivation
+
+Query Params currently require a bit of ceremony to be used at all. If Query
+Params today opted you out of the current restrictive-query-params implementation
+then the default experience with query params would "just work", improving the
+learning experience as well for new developers once they learn about `transitionTo`.
+
+
+## Detailed design
+
+This may require an optional feature to be non-breaking for folks that rely on
+forbidding all query params by default.
+
+Query Params as they work today are allow-list only, and block all other Query Params.
+
+Example:
+```js
+export default class ArticlesController extends Controller {
+  queryParams = ['category'];
+}
+```
+
+With `queryParams` specified, only `category` will be allowed on the route for
+this controller.
+
+This RFC only proposes a change only to this scenario:
+```js
+// Empty or default controller
+export default class ArticlesController extends Controller {
+}
+```
+
+where now `<RouterService>#transitionTo({ queryParams: { category: 'anything' } })`
+and LinkTo will _not_ strip away the `category` Query Param.
+
+The way this would work is that _all_ Query Params detected during a transition
+would live on the `ApplicationController`, as if the `queryParams` array somehow
+knew of all possible Query Params.
```

Comment:
Would this include adding a property with the same name as query parameter to application route when a query parameter is first used? Wouldn't this introduce the risk of collisions with existing properties or methods on application route? Imagine a query parameter `model`. 💥

What is the motivation behind this? Wouldn't it be much simpler if these query parameters are not coupled with controllers at all?

Reply:
> Wouldn't this introduce the risk of collisions with existing properties or methods on application route? Imagine a query parameter model.

That's a good call out. We'd need to add assertions for collisions of any existing property/method.

> Wouldn't it be much simpler if these query parameters are not coupled with controllers at all?

Conceptually, yes, but that's just how the implementation is today. This RFC aims to maximize ergonomics while minimizing code changes.

It _may_ be possible to implement this RFC by just saying that "if you're using implicit query params, they do not exist on any controller" -- which, maybe that's simpler, conceptually? 🤔

Reply:
> It may be possible to implement this RFC by just saying that "if you're using implicit query params, they do not exist on any controller" -- which, maybe that's simpler, conceptually? 🤔

I would prefer this. Less magic.